File: store/404.左叶子之和.js
/*
 * @lc app=leetcode.cn id=404 lang=javascript
 *
 * [404] 左叶子之和
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
let sum;
const getLeftSum = (node) => {
  // 前序
  // if (node.left && !node.left.left && !node.left.right) sum += node.left.val;
  // if (node.left) getLeftSum(node.left);
  // if (node.right) getLeftSum(node.right);
  // 中序
  if (!node) return 0;
  let leftSum = getLeftSum(node.left);
  if (node.left && !node.left.left && !node.left.right)
    leftSum += node.left.val;
  let rightSum = getLeftSum(node.right);
  return leftSum + rightSum;
};
var sumOfLeftLeaves = function (root) {
  // sum = 0;
  // getLeftSum(root);
  // return sum;

  return getLeftSum(root);
};
// @lc code=end

File: store/404.左叶子之和.c
/*
 * @lc app=leetcode.cn id=404 lang=c
 *
 * [404] 左叶子之和
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int num;

void back(struct TreeNode *root, int i)
{
  if (i == 0 && root->left == NULL && root->right == NULL)
  {
    num += root->val;
    return;
  }
  if (root->left != NULL)
    back(root->left, 0);
  if (root->right != NULL)
    back(root->right, 1);
}

int sumOfLeftLeaves(struct TreeNode *root)
{
  // 法一：递归
  // num = 0;
  // if (root != NULL)
  //   back(root, 1);
  // return num;
  // if (root == NULL)
  //   return 0;
  // if (root->left == NULL && root->right == NULL)
  //   return 0;
  // int leftVal = sumOfLeftLeaves(root->left); // 左孩子的左子节点
  // if (root->left != NULL && root->left->left == NULL && root->left->right == NULL)
  //   leftVal = root->left->val; // 左孩子为叶子节点
  // int rightVal = sumOfLeftLeaves(root->right);
  // return leftVal + rightVal;
  // 法二：迭代
  struct TreeNode **stack, *temp;
  int top;
  stack = (struct TreeNode **)malloc(sizeof(struct TreeNode *) * 100);
  num = top = 0;
  if (root != NULL)
  {
    stack[top++] = root;
    while (top > 0)
    {
      temp = stack[--top];
      if (temp->left != NULL && temp->left->left == NULL && temp->left->right == NULL)
        num += temp->left->val;
      if (temp->left != NULL)
        stack[top++] = temp->left;
      if (temp->right != NULL)
        stack[top++] = temp->right;
    }
  }
  return num;
}
// @lc code=end
